# Project 1 :- Bulid Decision Tree in which DV : Survived , IDV : Age,Sex&Fare & Predict the test dataset

In [47]:
import pandas as pd
import numpy as np
from sklearn import tree
from sklearn import preprocessing

In [48]:
titanic_train = pd.read_csv("H:/Shubham Coding/Classwork/Day 21 to/Day 24/train.csv")
titanic_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [49]:
titanic_train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [50]:
label_encoder = preprocessing.LabelEncoder()
titanic_train["Sex"] = label_encoder.fit_transform(titanic_train["Sex"])

In [51]:
predictors = pd.DataFrame([titanic_train["Sex"],titanic_train["Age"],titanic_train["Fare"]]).T

In [52]:
tree_model = tree.DecisionTreeClassifier(max_depth=6)

In [53]:
tree_model.fit(X=predictors,y=titanic_train["Survived"])

DecisionTreeClassifier(max_depth=6)

In [54]:
with open("Dtree1.dot",'w') as f:
    f =  tree.export_graphviz(tree_model,feature_names=["Sex","Age","Fare"],out_file=f);

In [57]:
tree_model.score(X=predictors,y=titanic_train["Survived"])

0.8301462317210349

# __The tree_model has given 83.01% Accuracy__

__tree_model Prediction :__

In [58]:
titanic_test = pd.read_csv("H:/Shubham Coding/Classwork/Day 21 to/Day 24/test.csv")
titanic_test.isnull().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [59]:
encoded_sex_test = label_encoder.fit_transform(titanic_test["Sex"])

In [61]:
test_features = pd.DataFrame([encoded_sex_test,titanic_test["Age"],titanic_test["Fare"]]).T

In [62]:
test_preds = tree_model.predict(X=test_features)
predicted_Output = pd.DataFrame({"PassengerId":titanic_test["PassengerId"],"Survived":test_preds})

In [63]:
predicted_Output.to_csv("Output.csv",index=False);

# The Given Test Data has been stored in Output.csv file

# Project 2 :- Attrition Rate Analysis in which DV : Attrition , IDV : Output of Random Forest Algorithm

In [17]:
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier

In [18]:
data = pd.read_csv("H:/Shubham Coding/Classwork/Day 21 to/Day 24/general_data.csv")
data.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [19]:
data = data.drop(columns = ['EmployeeID', 'EmployeeCount', 'Over18', 'StandardHours'])

In [20]:
data.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
Gender                      object
JobLevel                     int64
JobRole                     object
MaritalStatus               object
MonthlyIncome                int64
NumCompaniesWorked         float64
PercentSalaryHike            int64
StockOptionLevel             int64
TotalWorkingYears          float64
TrainingTimesLastYear        int64
YearsAtCompany               int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

In [21]:
data.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
PercentSalaryHike           0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [22]:
data["NumCompaniesWorked"].mean()

2.6948303347756775

In [23]:
data["TotalWorkingYears"].mean()

11.279936378095888

In [24]:
New_NumCompaniesWorked = np.where(data["NumCompaniesWorked"].isnull(),3,data["NumCompaniesWorked"])
New_TotalWorkingYears = np.where(data["TotalWorkingYears"].isnull(),11,data["TotalWorkingYears"])

In [25]:
data["NumCompaniesWorked"] = New_NumCompaniesWorked
data["TotalWorkingYears"] = New_TotalWorkingYears

In [26]:
objects = data.select_dtypes(include = "object").columns

In [27]:
label_encoder = preprocessing.LabelEncoder()
encoded = data.apply(label_encoder.fit_transform)

In [33]:
rf_model = RandomForestClassifier(n_estimators=1000,max_features =2,oob_score=True)

In [34]:
features =["Age","BusinessTravel", "Department", "DistanceFromHome", "Education", "EducationField", "Gender", "JobLevel", "JobRole",  "MaritalStatus", "MonthlyIncome", "NumCompaniesWorked",  "PercentSalaryHike", "StockOptionLevel",  "TotalWorkingYears", "TrainingTimesLastYear", "YearsAtCompany","YearsSinceLastPromotion", "YearsWithCurrManager"]

In [35]:
rf_model.fit(X=encoded[features],y=encoded["Attrition"])

RandomForestClassifier(max_features=2, n_estimators=1000, oob_score=True)

In [38]:
print("OOB Accuracy :",rf_model.oob_score_)

OOB Accuracy : 1.0


In [39]:
for feature, imp in zip(features,rf_model.feature_importances_):
    print(feature,imp)

Age 0.09697497566615801
BusinessTravel 0.028058649426942698
Department 0.02604458378167324
DistanceFromHome 0.06914900483769396
Education 0.041461951943638435
EducationField 0.041294932437051436
Gender 0.0184808527352395
JobLevel 0.037689619945958555
JobRole 0.05626608848185786
MaritalStatus 0.039718546587588395
MonthlyIncome 0.09422026904211285
NumCompaniesWorked 0.05489160796454582
PercentSalaryHike 0.06542376760768055
StockOptionLevel 0.0344974045506573
TotalWorkingYears 0.08480555763081317
TrainingTimesLastYear 0.045410515398879234
YearsAtCompany 0.06834853137570347
YearsSinceLastPromotion 0.04292676128135058
YearsWithCurrManager 0.05433637930445495


__As per Random Forest Algorithm & Random score value we will take all variables because oob score is 100%__

In [41]:
tree_model = tree.DecisionTreeClassifier(max_depth=8)

In [42]:
predictors =pd.DataFrame([encoded['Age'],encoded['MonthlyIncome'],encoded['TotalWorkingYears'],encoded['DistanceFromHome'],encoded['YearsAtCompany'],encoded['PercentSalaryHike'],encoded['NumCompaniesWorked'],encoded['JobRole'],encoded['YearsWithCurrManager']]).T

In [43]:
tree_model.fit(X=predictors,y=encoded["Attrition"])

DecisionTreeClassifier(max_depth=8)

In [44]:
with open("Attrition_Model.dot",'w') as f:
    f=tree.export_graphviz(tree_model,feature_names=["Age","MonthlyIncome","TotalWorkingYears","DistanceFromHome","YearsAtCompany","PercentSalaryHike","NumCompaniesWorked","JobRole","YearsWithCurrManager"],out_file=f);

In [45]:
print("Classification Accuracy with the features  is :", tree_model.score(X=predictors,y=encoded["Attrition"]))

Classification Accuracy with the features  is : 0.9018140589569161


# The Given Data has been stored in Attrition Model dot File

# Project 3 :- Build Decision Tree for Bank Loan Modelling DV - Personal Loan & IDV - Output of RF Model

In [46]:
import numpy as np
import pandas as pd
from sklearn import preprocessing
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier

In [47]:
data=pd.read_excel("H:/Shubham Coding/Classwork/Day 21 to/Day 24/Bank_Personal_Loan_Modelling.xlsx",sheet_name="Data")
data.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [48]:
data.drop(columns=["ID","ZIP Code",],inplace=True)

In [49]:
data.isnull().sum()

Age                   0
Experience            0
Income                0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [50]:
objects=data.select_dtypes(include="object").columns

In [51]:
label_encoder = preprocessing.LabelEncoder()
encoded = data.apply(label_encoder.fit_transform)

In [52]:
rf_model = RandomForestClassifier(n_estimators=1000,max_features =2,oob_score=True)

In [53]:
features =["Age", "Experience", "Income", "Family", "CCAvg", "Education","Mortgage", "Securities Account", "CD Account", "Online", "CreditCard"]

In [54]:
rf_model.fit(X=encoded[features],y=encoded["Personal Loan"])

RandomForestClassifier(max_features=2, n_estimators=1000, oob_score=True)

In [55]:
print("OOB Accuracy :",rf_model.oob_score_)

OOB Accuracy : 0.9874


__OOB Accuracy is 98.72%__

In [56]:
for feature, imp in zip(features,rf_model.feature_importances_):
    print(feature,imp)

Age 0.04508091016601001
Experience 0.04446245444274023
Income 0.3422540548877449
Family 0.0974768914201776
CCAvg 0.186391020255372
Education 0.16291732926638494
Mortgage 0.042524670548476105
Securities Account 0.005454047319619833
CD Account 0.054811061536275456
Online 0.00844722399906279
CreditCard 0.010180336158136146


In [57]:
tree_model =tree.DecisionTreeClassifier(max_depth = 8)

In [58]:
predictors=pd.DataFrame([data["Income"],data["CCAvg"],data["Education"],data["Family"]]).T

In [59]:
tree_model.fit(X=predictors,y=encoded["Personal Loan"])

DecisionTreeClassifier(max_depth=8)

In [60]:
with open("Bank_Model.dot",'w') as f:
    f=tree.export_graphviz(tree_model,feature_names=["Income","CCAvg","Education","Family"],out_file=f);

In [61]:
print("Classification Accuracy with the features  is :", tree_model.score(X=predictors,y=encoded["Personal Loan"]))

Classification Accuracy with the features  is : 0.9956


# Model Accuracy is 99.56%

# The Given Data is Stored in Bank Model dot File